[PM-18042] Build request response structure

coroiu opened the pull request:
## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->

## 📔 Objective

<!-- Describe what the purpose of this PR is, for example what bug you're fixing or new feature you're adding. -->

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

coroiu committed:
feat: scaffold rpc structure

coroiu committed:
wip: implement rpc request

coroiu committed:
feat: add async-trait to workspace and ipc

coroiu committed:
wip: scaffold all types for the handler registry

coroiu committed:
feat: let error be a handler implementation detail

coroiu committed:
feat: implement handler registry

coroiu committed:
feat: consolidate RpcPayload trait type

coroiu committed:
refactor: rename RpcRequest -> RpcRequestMessage

coroiu committed:
refactor: rename RpcResponse -> RpcResponseMessage

coroiu committed:
refactor: rename RpcPayload -> RpcRequest

coroiu committed:
refactor: rename payload -> request

coroiu committed:
wip: implement client rpc handling

coroiu committed:
feat: use RPC as a typed message

coroiu committed:
wip: process rpc request

coroiu committed:
feat: implement request handling

coroiu committed:
feat: implement discover command

coroiu committed:
feat: simplify discover response

coroiu committed:
fix: discover not working

coroiu committed:
fix: broken tests

coroiu committed:
fix: linting and docs

coroiu committed:
fix: lint

coroiu committed:
fix: typescript type showing up as `any`

coroiu committed:
feat: add abort signal support

coroiu commented on crates/bitwarden-ipc/src/ipc_client.rs:
```diff
@@ -186,6 +219,23 @@ where
+        <H::Request as TryInto<Vec<u8>>>::Error: std::fmt::Display,
+        <<H::Request as RpcRequest>::Response as TryFrom<Vec<u8>>>::Error: std::fmt::Display,
+        <<H::Request as RpcRequest>::Response as TryInto<Vec<u8>>>::Error: std::fmt::Display,
+    {
```
thought: I'm still very divided on this, maybe we should create a follow-up ticket where we simply hardcode the wire-format conversion

dani-garcia replied:
Question: What if we change `RpcRequest` to require Serialize and DeserializeOwned?
```rust
pub trait RpcRequest: serde::ser::Serialize + serde::de::DeserializeOwned {
    type Response: serde::ser::Serialize + serde::de::DeserializeOwned;

    /// Used to identify handlers
    fn name() -> String;
}
```

This would avoid a lot of the generic type bounds on all the methods, and I feel like most messages are going to just depend on serde to serialize/deserialize the data anyway. That would also centralize the wire-format to the IPC client, and we can swap to another one as long as it supports serde, which most formats do.

coroiu replied:
Lol VS Code renders your comment with the german flag 
![image](https://github.com/user-attachments/assets/178bea26-c32d-416e-aeb1-e37dc2894224)

I'll rewrite this for RPC and then we can follow-up with a similar rewrite for the top-level message structs

dani-garcia replied:
That's hilarious 😆

coroiu commented on crates/bitwarden-ipc/src/rpc/request_message.rs:
```diff
@@ -0,0 +1,32 @@
+    fn try_from(value: RpcRequestMessage) -> Result<Self, Self::Error> {
+        serde_json::to_vec(&value)
+    }
+}
```
thought: this is another thing I'm a bit divided on; the levels of serialization. At this point we don't know how to deserialize the request (because we don't know its type) but we still need to wrap it somehow and so we end up with a nested `Vec<u8>` field which contains the actual request data. This will then be deserialized by the rpc handler

coroiu replied:
Technically RPC can be seen as built on top of the req/res paradigm that the IPC client provides at its core, that's why we are actually defining a serilalization format (`serde_json`) here, because it's RPC over `TypedMessages`

coroiu committed:
chore: remove println

coroiu committed:
fix: try fixing type issue by running npm ci

coroiu committed:
fix: missing types during typecheck

dani-garcia commented on crates/bitwarden-ipc/src/wasm/error.rs:
```diff
@@ -0,0 +1,38 @@
+
+#[derive(Debug, Error, PartialEq, Eq)]
+#[bitwarden_error(flat, export_as = "RequestError")]
+pub enum JsRequestError {
```
Is this error duplicated because the original has the generic type? If so, I wonder if it won't be better to just remove the generic and use `Send(Box<dyn Debug>)` or `Send(String)` directly on the main type.

coroiu replied:
Yeah you're probably right, I did it with the other types because if how they were jumping across threads and `JsValue` couldn't do that anyways, but this one was still executed on the same thread. But should probably just do it with this one too for consistency

coroiu replied:
Fixed 👍

dani-garcia commented on crates/bitwarden-ipc/src/wasm/ipc_client.rs:
```diff
@@ -15,11 +15,16 @@ use crate::{
+    /// that interact with the IPC client, e.g. to register RPC handlers, trigger RPC requests,
+    /// send typed messages, etc. For examples see
+    /// [wasm::ipc_register_discover_handler](crate::wasm::ipc_register_discover_handler).
+    pub client: Arc<
```
This should probably be pub(crate), and then we don't need the skip

coroiu replied:
But then other teams won't be able to make their own functions using this?

dani-garcia replied:
You mean to let them run `register_rpc_handler` in their crates? Yeah that's fair, I didn't think of that

coroiu committed:
feat: simplify request error type

coroiu committed:
fix: lint

coroiu committed:
Refactor: Use serde instead of TryFrom<Vec<u8>> traits in IPC (#316)

## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe /
point to where this change is coming from. -->

## 📔 Objective

Refactoring the IPC to use serde instead. Putting this in a separate PR
to make reviewing easier

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or
informed the documentation
  team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire!
-->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry
that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or
concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or
indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

dani-garcia commented on crates/bitwarden-ipc/Cargo.toml:
```diff
@@ -20,15 +20,18 @@ wasm = [
+async-trait = { workspace = true }
 bitwarden-error = { workspace = true }
 bitwarden-threading = { workspace = true }
+erased-serde = "0.4.6"
```
```suggestion
erased-serde = ">=0.4.6, <0.5"
```

dani-garcia commented on crates/bitwarden-ipc/src/rpc/request.rs:
```diff
@@ -0,0 +1,8 @@
+    type Response: Serialize + DeserializeOwned + 'static;
+
+    /// Used to identify handlers
+    fn name() -> String;
```
For some other traits, instead of a function we use an associated constant, like on `PayloadTypeName`, should we do the same here?

coroiu replied:
Ah, yes, I only recently discovered that you could use constants like this (during a review I think) so I just missed updating this :)

dani-garcia commented on crates/bitwarden-ipc/src/rpc/request_message.rs:
```diff
@@ -0,0 +1,59 @@
+        &self.partial.request_type
+    }
+
+    pub fn full<T>(&self) -> Result<RpcRequestMessage<T>, RpcError>
```
`full` is a bit of a strange name for a function, maybe `deserialize_full` to be more clear?

dani-garcia reviewed (commented):
Some minor comments now that we have everything together, the rest looks pretty good to me!

coroiu committed:
fix: `erased-serde` version range

coroiu committed:
feat: use trait constant

coroiu committed:
feat: clarify function name

coroiu committed:
OMG THIS COMMA!!!

coroiu committed:
Merge branch 'main' into PM-18042-build-request-response-structure
